src/libfetchers/tarball.cc

```diff
@@ -225,6 +234,42 @@ struct TarballInputScheme : InputScheme
         return true;
     }
 
+};
+
+struct FileInputScheme : CurlInputScheme
+{
+    const std::string inputType() const override { return "file"; }
+
+    bool isValidURL(const ParsedURL & url) const override
+    {
+        auto parsedUrlScheme = parseUrlScheme(url.scheme);
+        return transportUrlSchemes.count(std::string(parsedUrlScheme.transport))
+            && (parsedUrlScheme.application
+                    ? parsedUrlScheme.application.value() == inputType()
+                    : !hasTarballExtension(url.path));
+    }
+
+    std::pair<StorePath, Input> fetch(ref<Store> store, const Input & input) override
+    {
+        auto file = downloadFile(store, getStrAttr(input.attrs, "url"), input.getName(), false);
```

Comment:
A problem with the use of `downloadFile()` is that it uses `FileIngestionMethod::Flat` instead of `FileIngestionMethod::Recursive`. Currently it's assumed that all flake inputs use recursive+sha256 with a name of "source". This allows inputs to be substituted using the `narHash` attribute in the lock file (see `Input::computeStorePath()`). However, the lazy trees branch will probably remove the ability to substitute inputs anyway...

Reply:
Oh indeed, I didn’t notice the `Flat` here. Would it be an issue to make `downloadFile` use `Recursive`? Afaik none of the other use sites really care about how the result is hashed since it’s only used internally by the fetchers but doesn’t leak outside.

(Alternatively I could just wrap that by adding an extra step that would copy the output to a `Recursive` CA location, but if that can be avoided it’s even better)